community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArrayHashMap.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2002-2018 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime;
+
+import org.opencypher.v9_0.util.InternalException;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import org.neo4j.helpers.collection.Pair;
+import org.neo4j.helpers.collection.PrefetchingIterator;
+
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.CONTINUE_PROBING;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.NOT_IN_USE;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.SLOT_EMPTY;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.VALUE_FOUND;
+
+/**
+ * A fast implementation of a hash map with long[] as keys.
+ */
+public class LongArrayHashMap<VALUE>
+{
+    private final int width;
+    private LongArrayHashTable table;
+    private Object[] values;
+
+    public LongArrayHashMap( int initialCapacity, int width )
+    {
+        assert (initialCapacity & (initialCapacity - 1)) == 0 : "Size must be a power of 2";
```

Comment:
"Capacity must be a power of 2" ? 

Like the bit-fiddling for checking power of 2!